Yul fuzzer improvements

bshastry opened the pull request:
Changes how function calls are synthesized (making them appear more often).

bshastry committed:
Changes to how function calls are generated.

bshastry committed:
Correct use of emplace_back

bshastry committed:
Changes to yul proto mutator.

leonardoalt commented on test/tools/ossfuzz/protomutators/YulProtoMutator.cpp:
```diff
@@ -107,9 +107,7 @@ template <typename T>
 T YPM::EnumTypeConverter<T>::validEnum(unsigned _seed)
 {
 	auto ret = static_cast<T>(_seed % (enumMax() - enumMin() + 1) + enumMin());
-	if constexpr (std::is_same_v<std::decay_t<T>, FunctionCall_Returns>)
```
It looks like `FunctionCall` "things" were removed, how does it work now?

bshastry replied:
Actually, a portion of the `FunctionCall` specification was altered. Earlier, the `FunctionCall` spec included the number of returned values and a pseudo-random index to select a previously declared function. The encoded logic was that, the proto converter selects a random (previously declared) function from the list of functions.

Say, the function signatures of previously declared functions are as follows:

```
f(a, b, c) -> d, e
g(a) -> b
h()
```

Now, the old spec worked like so. The fuzzer supplies
  - a random index say `11345` to select a function (we perform a modulo of number of previously declared functions)
  - the number of returned values desired (e.g., 0, 1, 2,3, 4 categorised as no return, single return, and multi return)

The converter would first perform a modulo on the index `11345 % 3 = 2` and select `h()` (see list of previously declared functions). In the next step, the converter would check if the function signature of the selected function matches the desired number of return values (no return, single, multiple).

In this case, although the converter found a function `f()` that could be used to make a function call, it may not have been inserted in the test case if the signature of the function did not match the number of return values specified by the fuzzer e.g., 1, 2, 3, 4.

I realised that this logic meant that fewer function calls are inserted into test cases generated by the fuzzer. Since there was no need to have the fuzzer additional specific number of return values of a function, I removed it to have a greater incidence of function calls.

I think a greater probability of function calls is useful to test optimiser steps like function inliner, and in general steps that simplify control logic.

In this PR, I removed the function selector index and number of function returns from the function call specification. Instead, the random selection is based on a monotonically increasing counter that already existed. Furthermore, once the function is selected, it is called in the generated test case.

leonardoalt replied:
Ok, thanks!

leonardoalt commented on test/tools/ossfuzz/protoToYul.h:
```diff
@@ -102,7 +102,15 @@ class ProtoConverter
 	void visit(SelfDestructStmt const&);
 	void visit(TerminatingStmt const&);
-	void visit(FunctionCall const&);
+	/// @param _f is the function call to be visited. @param _name is
```
Maybe keep one param per line?

bshastry replied:
Done.

leonardoalt reviewed (commented):
There's a lot of code removed, and I don't really see how the functionality is the same/similar. Can you explain a bit?

bshastry commented on test/tools/ossfuzz/yulProto.proto:
```diff
@@ -58,23 +58,14 @@ message Create {
+  required VarRef out_param1 = 5;
+  required VarRef out_param2 = 6;
+  required VarRef out_param3 = 7;
+  required VarRef out_param4 = 8;
```
The change here is that `Returns` and function index have been removed. In and out parameters remain but with different offsets within the message (unimportant detail).

bshastry committed:
Remove output parameters from function call since they are generated on the fly in the converter.

bshastry commented:
Thanks for the review @leonardoalt 

I rebased the PR and added a tiny update (remove out_params from function call spec). This is no longer needed (but I forgot to remove them). The diff beyond what you have already reviewed are four lines in `yulProto.proto` that define output parameters for a function call.

These have been removed because the converter creates a multi declaration statement for a function call assignment, like so

```
let x_1, ..., x_n = function_that_returns_n_values(i_1, ..., i_n)
```

The inputs (upto 4 parameters are supported) remain as they are. It's only the output parameters that are not needed.
